sys/net/gnrc/link_layer/netdev2/gnrc_netdev2_ieee802154.c

```diff
@@ -0,0 +1,206 @@
+            }
+            hdr = netif_hdr->data;
+            data = pkt->data;
+            hdr->lqi = data[nread + _LQI_TAIL_POS];
+            hdr->rssi = data[nread + _RSSI_TAIL_POS];
+            hdr->if_pid = thread_getpid();
+            pkt->type = dev->proto;
+#if ENABLE_DEBUG
+            DEBUG("_recv_ieee802154: received packet from %s of length %u\n",
+                  gnrc_netif_addr_to_str(src_str, sizeof(src_str),
+                                         gnrc_netif_hdr_get_src_addr(hdr),
+                                         hdr->src_l2addr_len),
+                  nread);
+#if defined(MODULE_OD)
+            od_hex_dump(pkt->data, nread, OD_WIDTH_DEFAULT);
+#endif
+#endif
+            gnrc_pktbuf_remove_snip(pkt, ieee802154_hdr);
+            LL_APPEND(pkt, netif_hdr);
+        }
+
+        DEBUG("_recv_ieee802154: reallocating.\n");
+        gnrc_pktbuf_realloc_data(pkt, nread);
+    }
+
+    return pkt;
+}
+
+static int _send(gnrc_netdev2_t *gnrc_netdev2, gnrc_pktsnip_t *pkt)
+{
+    netdev2_ieee802154_t *dev = (netdev2_ieee802154_t *)gnrc_netdev2->dev;
+    gnrc_netif_hdr_t *netif_hdr;
+    struct iovec *vector;
+    uint8_t *src, *dst = NULL;
+    int res = 0;
+    size_t n, src_len;
+    uint8_t mhr[IEEE802154_MAX_HDR_LEN];
+    uint8_t flags = (uint8_t)(dev->flags & NETDEV2_IEEE802154_SEND_MASK);
+    le_uint16_t dev_pan = byteorder_btols(byteorder_htons(dev->pan));
+
+    flags |= IEEE802154_FCF_TYPE_DATA;
+    if (pkt == NULL) {
+        DEBUG("_send_ieee802154: pkt was NULL\n");
+        return -EINVAL;
+    }
+    if (pkt->type != GNRC_NETTYPE_NETIF) {
+        DEBUG("_send_ieee802154: first header is not generic netif header\n");
+        return -EBADMSG;
+    }
+    netif_hdr = pkt->data;
+    /* prepare destination address */
+    if (netif_hdr->flags & /* If any of these flags is set so this is correct */
+        (GNRC_NETIF_HDR_FLAGS_BROADCAST | GNRC_NETIF_HDR_FLAGS_MULTICAST)) {
+        flags |= IEEE802154_BCAST;
+    }
+    else {
+        dst = gnrc_netif_hdr_get_dst_addr(netif_hdr);
+    }
+    src_len = netif_hdr->src_l2addr_len;
+    if (src_len > 0) {
+        src = gnrc_netif_hdr_get_src_addr(netif_hdr);
+    }
+    else if (dev->flags & NETDEV2_IEEE802154_SRC_MODE_LONG) {
+        src_len = IEEE802154_LONG_ADDRESS_LEN;
+        src = dev->long_addr;
+    }
+    else {
+        src_len = IEEE802154_SHORT_ADDRESS_LEN;
+        src = dev->short_addr;
+    }
+    /* fill MAC header, seq should be set by device */
+    if ((res = ieee802154_set_frame_hdr(mhr, src, src_len,
+                                        dst, netif_hdr->dst_l2addr_len,
+                                        dev_pan, dev_pan, flags,
+                                        dev->seq++)) == 0) {
+        DEBUG("_send_ieee802154: Error preperaring frame\n");
+        return -EINVAL;
+    }
+    /* prepare packet for sending */
+    pkt = gnrc_pktbuf_get_iovec(pkt, &n);
+    vector = (struct iovec *)pkt->data;
```

Comment:
I just copied that from `gnrc_netdev2_eth` and since we don't have any (or at least non that I know of that would be SO significant) memory leaks in on native or other Ethernet platforms, I guess it isn't duplicated.